# Carga de Librerias

In [2]:
import pandas as pd

# Carga del DataFrame

In [3]:
df = pd.read_csv('Amazon-Products.csv')

# DataFrame

In [4]:
df

Unnamed: 0                                               name  \
0                0  Lloyd 1.5 Ton 3 Star Inverter Split Ac (5 In 1...   
1                1  LG 1.5 Ton 5 Star AI DUAL Inverter Split AC (C...   
2                2  LG 1 Ton 4 Star Ai Dual Inverter Split Ac (Cop...   
3                3  LG 1.5 Ton 3 Star AI DUAL Inverter Split AC (C...   
4                4  Carrier 1.5 Ton 3 Star Inverter Split AC (Copp...   
...            ...                                                ...   
551580        1099                Adidas Regular Fit Men's Track Tops   
551581        1100          Redwolf Noice Toit Smort - Hoodie (Black)   
551582        1101     Redwolf Schrute Farms B&B - Hoodie (Navy Blue)   
551583        1102                                    Puma Men Shorts   
551584        1103  Mothercare Printed Cotton Elastane Girls Infan...   

           main_category      sub_category  \
0             appliances  Air Conditioners   
1             appliances  Air Conditioners   
2             appliances  Air Conditioners   
3             appliances  Air Conditioners   
4             appliances  Air Conditioners   
...                  ...               ...   
551580  sports & fitness              Yoga   
551581  sports & fitness              Yoga   
551582  sports & fitness              Yoga   
551583  sports & fitness              Yoga   
551584  sports & fitness              Yoga   

                                                    image  \
0       https://m.media-amazon.com/images/I/31UISB90sY...   
1       https://m.media-amazon.com/images/I/51JFb7FctD...   
2       https://m.media-amazon.com/images/I/51JFb7FctD...   
3       https://m.media-amazon.com/images/I/51JFb7FctD...   
4       https://m.media-amazon.com/images/I/41lrtqXPiW...   
...                                                   ...   
551580  https://m.media-amazon.com/images/I/71tHAR9pIY...   
551581  https://m.media-amazon.com/images/I/41pKrMZ5lQ...   
551582  https://m.media-amazon.com/images/I/41n9u+zNSc...   
551583  https://m.media-amazon.com/images/I/51LoWv5JDt...   
551584  https://m.media-amazon.com/images/I/814X-sI3DB...   

                                                     link ratings  \
0       https://www.amazon.in/Lloyd-Inverter-Convertib...     4.2   
1       https://www.amazon.in/LG-Convertible-Anti-Viru...     4.2   
2       https://www.amazon.in/LG-Inverter-Convertible-...     4.2   
3       https://www.amazon.in/LG-Convertible-Anti-Viru...     4.0   
4       https://www.amazon.in/Carrier-Inverter-Split-C...     4.1   
...                                                   ...     ...   
551580  https://www.amazon.in/Adidas-Regular-Mens-Trac...     3.2   
551581  https://www.amazon.in/Redwolf-Noice-Smort-Cott...     2.0   
551582  https://www.amazon.in/Redwolf-Schrute-Farms-Ho...     4.0   
551583  https://www.amazon.in/Puma-Woven-Short-5208526...     4.4   
551584  https://www.amazon.in/Mothercare-Carrot-Regula...     4.6   

       no_of_ratings discount_price actual_price  
0              2,255        ₹32,999      ₹58,990  
1              2,948        ₹46,490      ₹75,990  
2              1,206        ₹34,490      ₹61,990  
3                 69        ₹37,990      ₹68,990  
4                630        ₹34,490      ₹67,790  
...              ...            ...          ...  
551580             9         ₹3,449       ₹4,599  
551581             2         ₹1,199       ₹1,999  
551582             1         ₹1,199       ₹1,999  
551583            37            NaN          NaN  
551584             5         ₹1,039       ₹1,299  

[551585 rows x 10 columns]

Vemos que el DataFrame cuenta con mas de 500.000 datos, hay 10 columnas de las cuales la primera se llama Unnamed, pero es el indice, las otras 9 cuentan con nombres de los productos, categorias, subcategorias, ranking, cantidad de veces que se califico, imagen del producto, link del producto, descuento, precio.

Vamos a renombrar primero la columna `Unnamed`como `index`

In [5]:
df = df.rename(columns={"Unnamed: 0": "index"})

## Nulos

Ahora que ya se modifico el nombre de la columna veamos si en el dataframe cuenta con valores nulos

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 551585 entries, 0 to 551584
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   index           551585 non-null  int64 
 1   name            551585 non-null  object
 2   main_category   551585 non-null  object
 3   sub_category    551585 non-null  object
 4   image           551585 non-null  object
 5   link            551585 non-null  object
 6   ratings         375791 non-null  object
 7   no_of_ratings   375791 non-null  object
 8   discount_price  490422 non-null  object
 9   actual_price    533772 non-null  object
dtypes: int64(1), object(9)
memory usage: 42.1+ MB


### Ratings

Vemos que las columnas con `NaN` son `ratings, no_of_ratings, discount_price y actual_price`, vamos a rellenar cada una de ellas

In [7]:
df['ratings'].unique()

array(['4.2', '4.0', '4.1', '4.3', '3.9', '3.8', '3.5', nan, '4.6', '3.3',
       '3.4', '3.7', '2.9', '5.0', '4.4', '3.6', '2.7', '4.5', '3.0',
       '3.1', '3.2', '4.8', '4.7', '2.5', '1.0', '2.6', '2.8', '2.3',
       '1.7', 'Get', '1.8', '2.4', '4.9', '2.2', '1.6', '1.9', '2.0',
       '1.4', '2.1', 'FREE', '1.2', '1.3', '1.5', '₹68.99', '₹65', '1.1',
       '₹70', '₹100', '₹99', '₹2.99'], dtype=object)

Hay valores nulos (`NaN`), al igual que columnas con palabras, como `Get y FREE` y otras con precios en vez del valor del raiting, esto se debe a que la columna posee un `Dtype`de `object`, cuando deberia ser un número flotante, al cambiar del `dtype` a `float` a los valores que no son numericos los establece como `NaN` lo que nos ahorra el paso de tener que transformalos como nulos, la otra ventaja es que al ser valores flotantes se puede realizar opercaciones estadisticas y matematicas sobre ella.

In [8]:
df['ratings'] = pd.to_numeric(df['ratings'], errors='coerce')

In [9]:
df['ratings'].describe()

count    369558.000000
mean          3.832311
std           0.756102
min           1.000000
25%           3.500000
50%           3.900000
75%           4.300000
max           5.000000
Name: ratings, dtype: float64

In [10]:
df['ratings'].unique()

array([4.2, 4. , 4.1, 4.3, 3.9, 3.8, 3.5, nan, 4.6, 3.3, 3.4, 3.7, 2.9,
       5. , 4.4, 3.6, 2.7, 4.5, 3. , 3.1, 3.2, 4.8, 4.7, 2.5, 1. , 2.6,
       2.8, 2.3, 1.7, 1.8, 2.4, 4.9, 2.2, 1.6, 1.9, 2. , 1.4, 2.1, 1.2,
       1.3, 1.5, 1.1])

Relleno los valores nulos con la mediana ya que es casi lo mismo 3.83 que 3.9 y de esta forma nos evitamos tener varios números despues de la coma

In [11]:
mediana_ratings = df['ratings'].median()
df['ratings'] = df['ratings'].fillna(mediana_ratings)
df['ratings'].unique()

array([4.2, 4. , 4.1, 4.3, 3.9, 3.8, 3.5, 4.6, 3.3, 3.4, 3.7, 2.9, 5. ,
       4.4, 3.6, 2.7, 4.5, 3. , 3.1, 3.2, 4.8, 4.7, 2.5, 1. , 2.6, 2.8,
       2.3, 1.7, 1.8, 2.4, 4.9, 2.2, 1.6, 1.9, 2. , 1.4, 2.1, 1.2, 1.3,
       1.5, 1.1])

### No_of_ratings

In [12]:
df['no_of_ratings'].unique()

array(['2,255', '2,948', '1,206', ..., '3,329', '7,141', '4,406'],
      dtype=object)

Filtro el DataFrame para obtener solo las filas donde 'no_of_ratings' contiene letras

In [13]:
filas_con_palabras = df[df['no_of_ratings'].str.contains(r'[A-Za-z]', na=False)]
filas_con_palabras.head()

index                                               name main_category  \
437    437  Daikin 1.8 Ton 3 Star Inverter Split Ac (Coppe...    appliances   
473    473  Master Electronic provides Window AC 3-Star 1....    appliances   
474    474  Master Electronic provides 1.5 Ton 3 Star Dust...    appliances   
475    475  Master Electronics Provides 1 Ton 3 Star Inver...    appliances   
488    488            1.5 Ton 5 Star Inverter Split AC, White    appliances   

         sub_category                                              image  \
437  Air Conditioners  https://m.media-amazon.com/images/I/51hn7NHS3T...   
473  Air Conditioners  https://m.media-amazon.com/images/I/41B0LWORAr...   
474  Air Conditioners  https://m.media-amazon.com/images/I/61yVE5P7FU...   
475  Air Conditioners  https://m.media-amazon.com/images/I/11mMBc5fl2...   
488  Air Conditioners  https://m.media-amazon.com/images/I/11au-UO4sB...   

                                                  link  ratings  \
437  https://www.amazon.in/Daikin-Inverter-Copper-F...      3.9   
473  https://www.amazon.in/Master-Electronic-provid...      3.9   
474  https://www.amazon.in/Master-Electronic-provid...      3.9   
475  https://www.amazon.in/Master-Electronics-Provi...      3.9   
488  https://www.amazon.in/1-5-Star-Inverter-Split-...      3.9   

             no_of_ratings discount_price actual_price  
437  Only 2 left in stock.        ₹59,979      ₹72,900  
473  Only 1 left in stock.        ₹36,999      ₹39,999  
474  Only 1 left in stock.        ₹36,999      ₹39,666  
475  Only 1 left in stock.        ₹46,999      ₹49,666  
488  Only 2 left in stock.        ₹45,000      ₹50,000

Vemos que hay descripciones en algunas filas cuando deberian ser solo el número de veces que se califico dicho articulo, al cambiar el dtype de la columna se resolvera tambien este incombeniente


In [14]:
df['no_of_ratings'] = pd.to_numeric(df['no_of_ratings'], errors='coerce')
df['no_of_ratings'].describe()

count    339029.000000
mean         88.908810
std         171.958778
min           1.000000
25%           4.000000
50%          15.000000
75%          79.000000
max         999.000000
Name: no_of_ratings, dtype: float64

En este caso si voy a usar la med

In [15]:
media_no_ratings = df['no_of_ratings'].mean()
df['no_of_ratings'] = df['no_of_ratings'].fillna(media_no_ratings)
df['no_of_ratings'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 551585 entries, 0 to 551584
Series name: no_of_ratings
Non-Null Count   Dtype  
--------------   -----  
551585 non-null  float64
dtypes: float64(1)
memory usage: 4.2 MB


### Discount_price

Para esta columna a los valores nulos los voy a poner como 0 indicando que no tienen descuento

In [16]:
df['discount_price'] = df['discount_price'].fillna(0)
df['discount_price'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 551585 entries, 0 to 551584
Series name: discount_price
Non-Null Count   Dtype 
--------------   ----- 
551585 non-null  object
dtypes: object(1)
memory usage: 4.2+ MB


### Actual_price 

En esta parte voy a eliminar las filas con valores ausentes, porque no puedo saber cual era el valor del producto cuando se publico.

In [17]:
df = df.dropna(subset=['actual_price'])

In [18]:
df

index                                               name  \
0           0  Lloyd 1.5 Ton 3 Star Inverter Split Ac (5 In 1...   
1           1  LG 1.5 Ton 5 Star AI DUAL Inverter Split AC (C...   
2           2  LG 1 Ton 4 Star Ai Dual Inverter Split Ac (Cop...   
3           3  LG 1.5 Ton 3 Star AI DUAL Inverter Split AC (C...   
4           4  Carrier 1.5 Ton 3 Star Inverter Split AC (Copp...   
...       ...                                                ...   
551579   1098  BE SAVAGE Pink I Purple You Hoodie Suitable fo...   
551580   1099                Adidas Regular Fit Men's Track Tops   
551581   1100          Redwolf Noice Toit Smort - Hoodie (Black)   
551582   1101     Redwolf Schrute Farms B&B - Hoodie (Navy Blue)   
551584   1103  Mothercare Printed Cotton Elastane Girls Infan...   

           main_category      sub_category  \
0             appliances  Air Conditioners   
1             appliances  Air Conditioners   
2             appliances  Air Conditioners   
3             appliances  Air Conditioners   
4             appliances  Air Conditioners   
...                  ...               ...   
551579  sports & fitness              Yoga   
551580  sports & fitness              Yoga   
551581  sports & fitness              Yoga   
551582  sports & fitness              Yoga   
551584  sports & fitness              Yoga   

                                                    image  \
0       https://m.media-amazon.com/images/I/31UISB90sY...   
1       https://m.media-amazon.com/images/I/51JFb7FctD...   
2       https://m.media-amazon.com/images/I/51JFb7FctD...   
3       https://m.media-amazon.com/images/I/51JFb7FctD...   
4       https://m.media-amazon.com/images/I/41lrtqXPiW...   
...                                                   ...   
551579  https://m.media-amazon.com/images/I/61voQ68VuT...   
551580  https://m.media-amazon.com/images/I/71tHAR9pIY...   
551581  https://m.media-amazon.com/images/I/41pKrMZ5lQ...   
551582  https://m.media-amazon.com/images/I/41n9u+zNSc...   
551584  https://m.media-amazon.com/images/I/814X-sI3DB...   

                                                     link  ratings  \
0       https://www.amazon.in/Lloyd-Inverter-Convertib...      4.2   
1       https://www.amazon.in/LG-Convertible-Anti-Viru...      4.2   
2       https://www.amazon.in/LG-Inverter-Convertible-...      4.2   
3       https://www.amazon.in/LG-Convertible-Anti-Viru...      4.0   
4       https://www.amazon.in/Carrier-Inverter-Split-C...      4.1   
...                                                   ...      ...   
551579  https://www.amazon.in/Be-savage-Purple-Hoodie-...      4.6   
551580  https://www.amazon.in/Adidas-Regular-Mens-Trac...      3.2   
551581  https://www.amazon.in/Redwolf-Noice-Smort-Cott...      2.0   
551582  https://www.amazon.in/Redwolf-Schrute-Farms-Ho...      4.0   
551584  https://www.amazon.in/Mothercare-Carrot-Regula...      4.6   

        no_of_ratings discount_price actual_price  
0            88.90881        ₹32,999      ₹58,990  
1            88.90881        ₹46,490      ₹75,990  
2            88.90881        ₹34,490      ₹61,990  
3            69.00000        ₹37,990      ₹68,990  
4           630.00000        ₹34,490      ₹67,790  
...               ...            ...          ...  
551579        6.00000           ₹711         ₹999  
551580        9.00000         ₹3,449       ₹4,599  
551581        2.00000         ₹1,199       ₹1,999  
551582        1.00000         ₹1,199       ₹1,999  
551584        5.00000         ₹1,039       ₹1,299  

[533772 rows x 10 columns]

Compruebo que todas las filas estan correctas y sin nulos

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 533772 entries, 0 to 551584
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   index           533772 non-null  int64  
 1   name            533772 non-null  object 
 2   main_category   533772 non-null  object 
 3   sub_category    533772 non-null  object 
 4   image           533772 non-null  object 
 5   link            533772 non-null  object 
 6   ratings         533772 non-null  float64
 7   no_of_ratings   533772 non-null  float64
 8   discount_price  533772 non-null  object 
 9   actual_price    533772 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 44.8+ MB


In [21]:
df.to_csv("Amazon.csv", index=False)